### DISCUSSION TITLE
executeAsync should throw errors

### DISCUSSION THREAD

**Discussion by anarkrypto** [next-safe-action/next-safe-action#240] [Ideas]
It is intuitively expected that executeAsync will throw an error rather than resolving the promise with a  `{ serverError: "message" }` object.

Without rejection, the implementation with libs like **react-hook-form** does not produce the expected effects, as **formState.isSubmitSuccessful** will always be true even if it is not successful.

Consider react-query's **mutateAsync**:

> Use mutateAsync instead of mutate to get a promise which will resolve on success or **throw on an error**.
> This can for example be used to compose side effects.

https://tanstack.com/query/v4/docs/framework/react/guides/mutations#promises

**Comment by TheEdoRan** [comment]
Hmm, I understand your point, but calling the action directly doesn't throw errors either. Right now you can set `throwValidationErrors` and throw server errors inside `handleReturnedServerError` at the client level, and/or use `throwValidationErrors` and `throwServerError` at the action level to throw errors.

Then, for instance, how would you handle validation errors when using `executeAsync`? Would it be better to throw them like you suggest with server errors or return them (current implementation)?

**Comment by anarkrypto** [comment]
> calling the action directly doesn't throw errors either.

Of course, using react actions has numerous limitations. That's why we use next-safe-action.

> Then, for instance, how would you handle validation errors when using executeAsync? Would it be better to throw them like you suggest with server errors or return them (current implementation)?

I prefer to use **onError** callback. However executeAsync could return a custom extended Error just like **zod** does.

> Right now you can set throwValidationErrors and throw server errors inside handleReturnedServerError at the client level, and/or use throwValidationErrors and throwServerError at the action level to throw errors.

Great, throw server errors inside handleReturnedServerError at the client level solves the problem.

I believe the question now is whether it is worth making this a standard behavior in next-safe-action.  I believe it would be more intuitive.

**Reply by TheEdoRan** [reply]
> Of course, using react actions has numerous limitations. That's why we use next-safe-action.

Sorry, I've put it badly. I'm referring to direct execution of actions defined using next-safe-action, docs [here](https://next-safe-action.dev/docs/execution/direct-execution). Right now, directly calling them is essentially the same as using `executeAsync`. So my question was: should the behavior be the same or should `executeAsync` behave differently (by throwing server errors instead of returning them)? Hope it's clearer this time.

**Reply by anarkrypto** [reply]
@TheEdoRan I believe it also applies.

resolve (fulfilled state) means that the operation was completed successfully.
I don't think "success fulfill with error" makes sense

Certainly rejecting will avoid several false positive bugs like the one with **formState.isSubmitSuccessful**
I guess there is there several other devs experiencing false positives right now because of the current implementation

Rejection is the default and expected behavior for a promise with error.
The opposite (never rejection) would depend on good reasons, it's not intuitive and adds extra complexity for devs

**Reply by TheEdoRan** [reply]
I personally think the current default is good, because it catches all errors that happen on action execution and return them to the client with a generic string, which is usually what you want, since Server Actions are intended to be called from the client. But you can always customize that behavior, and `handleReturnedServerError` (at the client level) and `throwServerError` (at the action level) allows you to do that.

**Reply by anarkrypto** [reply]
@TheEdoRan what happened? now we don't have handleReturnedServerError and the handleServerError seems to execute directly on the server. I cannot throw inside it anymore to be able of throw the error with a message in the client side.

**Reply by TheEdoRan** [reply]
@anarkrypto [v7.9.0](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.9.0) - `handleServerError` has the same functionality as the previous `handleReturnedServerError`, in fact I've just updated its name. `handleServerErrorLog` was removed, since you can easily handle logging in the same function. What you return/throw from `handleServerError` is what you get as `serverError` on the client, just like you did with `handleReturnedServerError`.

> `handleServerError` seems to execute directly on the server

`handleReturnedServerError` always executed on the server too, the return value (or thrown error) is what you get on the client.

**Reply by anarkrypto** [reply]
@TheEdoRan but now I cannot produce side effects on the client side. If I throw in the server, I do not get the message. If I just return, it does not cause the throw to side effect.

So it's a false success

![image](https://github.com/user-attachments/assets/1a23dfc1-9313-4848-b41a-ca6d65977ef2)

**Reply by anarkrypto** [reply]
before I could make this to throw on the client side, producing the side effect

```bash
async handleReturnedServerError(error) {
		let message = 'Unknown Error'

		if (error instanceof Error) {
			message = error.message
		}

		throw new Error(message)
	}
	
```

**Reply by TheEdoRan** [reply]
@anarkrypto Same Next.js version as before? Can you please provide a reproduction link? Thanks.

**Reply by anarkrypto** [reply]
@TheEdoRan, not the same version, I am currently using the lastest next version. Do you think something changed in next to produce this? It's not the fact that handleReturnedServerError was run on the client side while handleServerError is on the server?

I do not have a minimal reproduction yet, but you can test it here: https://nanopayme-git-main-nanopay.vercel.app/login
Try to login with a fake credential. As you see, since I am throwing at server side, vercel does not return any message.

Here is the code: https://github.com/nanopay/nanopay.me/blob/bd780bcf8f2a1d2c6d23d886febe30ab1bbd5bba/src/lib/safe-action.ts#L5-L13

**Reply by TheEdoRan** [reply]
> not the same version, I am currently using the lastest next version.

If you updated packages outside of this library, the issue could be related to them. To find out what's causing problems, try to downgrade the other packages to their previous versions and see if the issue persists.

> It's not the fact that handleReturnedServerError was run on the client side while handleServerError is on the server?

`handleReturnedServerError` -> `handleServerError`. I've just updated the name, same functionality as before, always ran on the server, no difference!

I've just tried to throw the error inside `handleServerError` in the playground app of this repo and it works fine. Not sure if something changed with the latest Next.js versions, if you find the time please provide a minimal reproduction, thanks.

**Reply by anarkrypto** [reply]
@TheEdoRan here is a minimal reproduction test with next v14.2.14

Repo: https://github.com/anarkrypto/next-safe-action-cannot-throw
Demo: https://next-safe-action-cannot-throw.vercel.app

As you see, the error is only returned when the handleServerError return as a string, instead throw.
When we throw inside the server, next.js does not return a body.

**Reply by TheEdoRan** [reply]
@anarkrypto I've just tested your repro, it seems to work fine on my end. When I click on "throw error B" it throws the error correctly, and this gets logged in browser console:

![image](https://github.com/user-attachments/assets/045c38ea-2e44-4d7e-a81b-188a661fc268)

Unless I'm missing something, I don't see where the bug is.
